ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobMetadata.java

```diff
@@ -399,8 +420,72 @@ public boolean equals(Object o) {
       return false;
     }
     CloudBlobMetadata om = (CloudBlobMetadata) o;
-    return (Objects.equals(id, om.id) && Objects.equals(partitionId, om.partitionId) && creationTime == om.creationTime
+    return (Objects.equals(id, om.id) && Objects.equals(partitionId, om.partitionId) && accountId == om.accountId
+        && containerId == om.containerId && size == om.size && creationTime == om.creationTime
         && uploadTime == om.uploadTime && expirationTime == om.expirationTime && deletionTime == om.deletionTime
-        && size == om.size && accountId == om.accountId && containerId == om.containerId);
+        && namingScheme == om.namingScheme && encryptionOrigin == om.encryptionOrigin && Objects.equals(vcrKmsContext,
+        om.vcrKmsContext) && Objects.equals(cryptoAgentFactory, om.cryptoAgentFactory)
+        && encryptedSize == om.encryptedSize);
+  }
+
+  /**
+   * @return a {@link HashMap} of metadata key-value pairs.
+   */
+  public HashMap<String, String> toMap() {
+    ObjectMapper objectMapper = new ObjectMapper();
+    HashMap<String, String> map = new HashMap<>();
+    ObjectNode node = objectMapper.convertValue(this, ObjectNode.class);
+    node.fieldNames().forEachRemaining(fieldName -> {
+      map.put(fieldName, node.get(fieldName).asText());
+    });
+    return map;
+  }
+
+  /** Custom serializer for CloudBlobMetadata class that omits fields with default values. */
+  static class MetadataSerializer extends StdSerializer<CloudBlobMetadata> {
+
+    public MetadataSerializer() {
+      this(null);
+    }
+
+    public MetadataSerializer(Class<CloudBlobMetadata> t) {
+      super(t);
+    }
+
+    @Override
+    public void serialize(CloudBlobMetadata value, JsonGenerator gen, SerializerProvider provider) throws IOException {
+
+      gen.writeStartObject();
+      // Required fields
+      gen.writeStringField(FIELD_ID, value.id);
+      gen.writeStringField(FIELD_PARTITION_ID, value.partitionId);
+      gen.writeNumberField(FIELD_ACCOUNT_ID, value.accountId);
+      gen.writeNumberField(FIELD_CONTAINER_ID, value.containerId);
+      gen.writeNumberField(FIELD_SIZE, value.size);
+      // Optional fields with default values to exclude
+      if (value.creationTime > 0) {
+        gen.writeNumberField(FIELD_CREATION_TIME, value.creationTime);
+      }
+      if (value.uploadTime > 0) {
+        gen.writeNumberField(FIELD_UPLOAD_TIME, value.uploadTime);
+      }
+      if (value.deletionTime > 0) {
```

Comment:
should these time fields be >= 0, since technically 0 could be a valid timestamp?

Reply:
I don't see any practical use case for 0 to be a valid timestamp.

Reply:
That is true, unless a time traveler from 1970 uploads a blob. We can keep this as is